pyensign/ml/loader.py

```diff
@@ -0,0 +1,75 @@
+from typing import AsyncIterator, Iterator, List, Any
+
+from langchain_core.document_loaders import BaseLoader
+from langchain_core.documents import Document
+
+from pyensign.ensign import Ensign
+
+
+class EnsignLoader(BaseLoader):
+    """
+    A loader that reads data from an Ensign topic.
+    """
+
+    def __init__(self, topic: str, content_field: str = "", **kwargs):
+        """
+        Initialize the loader with a topic name. Loading data from Ensign requires a
+        topic name and Ensign credentials. Credentials can be passed in as keyword
+        arguments or set as the ENSIGN_CLIENT_ID and ENSIGN_CLIENT_SECRET environment
+        variables.
+
+        Args:
+            topic: The name of the topic to read from.
+        """
+
+        if len(topic) == 0:
+            raise ValueError("Topic name is required")
+
+        self.topic = topic
+        self.content_field = content_field
+        self.ensign = Ensign(**kwargs)
+
+    def _convert_to_document(self, event: Any) -> Document:
+        """
+        Convert an Ensign event to a Document.
+        """
+
+        return Document(
+            page_content=event.decode_to_str(text_field=self.content_field),
+            metadata=event.meta,
+        )
+
+    def load(self) -> Iterator[Document]:
+        """
+        Loads all documents from Ensign into memory.
+        Note: In production, this should be used with caution since it loads all
+        documents into memory at once.
+        """
+
+        raise NotImplementedError
+
+    async def aload(self) -> List[Document]:
```

Comment:
Since I'm subclassing the BaseLoader, I'm just using the methods that already exist for compatibility reasons: https://python.langchain.com/v0.1/docs/modules/data_connection/document_loaders/custom/. So unfortunately, I think we're tied to those names.